spring-integration-core/src/main/java/org/springframework/integration/aggregator/MessageTransformingAggregatingMessageGroupProcessor.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.aggregator;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+import org.springframework.integration.IntegrationMessageHeaderAccessor;
+import org.springframework.integration.store.MessageGroup;
+import org.springframework.integration.support.MessageBuilderFactory;
+import org.springframework.messaging.Message;
+
+/**
+ * This implementation of MessageGroupProcessor will take the messages from the MessageGroup
+ * and pass them on in a single message with a Collection of transformed Messages as a payload.
+ *<p>
+ * By default, this implementation does not transform the payload of the messages but simply removes
+ * {@link IntegrationMessageHeaderAccessor#SEQUENCE_SIZE}, {@link IntegrationMessageHeaderAccessor#SEQUENCE_NUMBER} and
+ * {@link IntegrationMessageHeaderAccessor#CORRELATION_ID}.
+ *
+ * @author Alen Turkovic
+ * @since 5.1
+ */
+public class MessageTransformingAggregatingMessageGroupProcessor extends AbstractAggregatingMessageGroupProcessor {
+
+	private BiFunction<Message<?>, MessageBuilderFactory, Message<?>> projection = (m, factory) -> factory
+			.fromMessage(m)
+			.removeHeaders(
```

Comment:
I'm a bit reluctant to introduce some transformation functionality into the correlation handler.
It's really not its responsibility. Plus I'm not sure why would one need to remove all these useful headers if you really are going to return a payload as a `List<Message<?>>`. You are going to iterate over them downstream anyway, so `IntegrationMessageHeaderAccessor.SEQUENCE_NUMBER` would be great indicator for each of them.

I can see only much simple approach to really build single `payload` from the `List<Message<?>>` and that's all:
````java
public class CollectionAggregatingMessageGroupProcessor implement MessageGroupProcessor {
    public Object processMessageGroup(MessageGroup group) {
              return MessageBuilder.withPayload(group.getMessages());
    }
}
````

All other hard transforming logic I would leave for the downstream flow...

Reply:
`IntegrationMessageHeaderAccessor.SEQUENCE_NUMBER` might be a mistake, you are right.

Transformation projection might have also been a mistake, initially, I have just removed `IntegrationMessageHeaderAccessor.SEQUENCE_SIZE` and `IntegrationMessageHeaderAccessor.CORRELATION_ID` since they are written in the aggregated message headers, and extracted the projection later which might have been a mistake...

Reply:
Well, with my idea you don't need any aggregation logic at all: all the information we need is present in each message in the payload. Isn't it?

Reply:
I'm not sure I understand the framework well enough so I am afraid I am not following you, sorry.

```
  @Bean
  public IntegrationFlow out() {
    return IntegrationFlows.from("channel-out")
      // a custom aggregating processor so we keep headers of each message
      // removed duplicated correlation headers
      // by default, we would lose the 'from' header
      .aggregate(a -> a.outputProcessor(new MessageTransformingAggregatingMessageGroupProcessor()))
      .log()
      .get();
  }
```
This is an example of how I used it for testing, the only reason I removed these headers is that they were (as I initially though, but might be wrong) unnecessarily duplicated for each message since they are also written in the aggregated result.

Reply:
Thanks for the thorough explanation, that helps clarify things a lot :)

I thought these headers were not necessary and added noise, but reading your explanation this is clearly not the case. 

Sorry, please close the PR.